src/Microsoft.Data/StringColumn.cs

```diff
@@ -273,26 +271,76 @@ private StringColumn Clone(PrimitiveColumn<bool> boolColumn)
+                PrimitiveColumn<long> longMapIndices = mapIndices as PrimitiveColumn<long>;
+                longMapIndices.ApplyElementwise((long? mapIndex, long rowIndex) =>
                 {
-                    ret[i] = this[getIndex(i).Value];
-                }
+                    long index = rowIndex;
+                    if (invertMapIndices)
+                        index = longMapIndices.Length - 1 - index;
+                    if (index < setBufferMinRange || index >= setBufferMaxRange)
+                    {
+                        int bufferIndex = (int)(index / maxCapacity);
+                        setBuffer = ret._stringBuffers[bufferIndex];
+                        setBufferMinRange = bufferIndex * maxCapacity;
+                        setBufferMaxRange = (bufferIndex + 1) * maxCapacity;
+                    }
+                    index -= setBufferMinRange;
+                    if (mapIndex == null)
+                    {
+                        setBuffer[(int)index] = null;
+                        ret._nullCount++;
+                        return mapIndex;
+                    }
+
+                    if (mapIndex.Value < getBufferMinRange || mapIndex.Value >= getBufferMaxRange)
+                    {
+                        int bufferIndex = (int)(mapIndex.Value / maxCapacity);
+                        getBuffer = _stringBuffers[bufferIndex];
+                        getBufferMinRange = bufferIndex * maxCapacity;
+                        getBufferMaxRange = (bufferIndex + 1) * maxCapacity;
+                    }
+                    int bufferLocalMapIndex = (int)(mapIndex - getBufferMinRange);
+                    setBuffer[(int)index] = getBuffer[bufferLocalMapIndex];
+                    return mapIndex;
+                });
             }
-            else
+            else if (mapIndices.DataType == typeof(int))
             {
-                long mapIndicesLengthIndex = mapIndices.Length - 1;
-                for (long i = 0; i < mapIndices.Length; i++)
+                PrimitiveColumn<int> intMapIndices = mapIndices as PrimitiveColumn<int>;
+                intMapIndices.ApplyElementwise((int? mapIndex, long rowIndex) =>
                 {
-                    ret[i] = this[getIndex(mapIndicesLengthIndex - i).Value];
-                }
+                    long index = rowIndex;
+                    if (invertMapIndices)
+                        index = intMapIndices.Length - 1 - index;
+
+                    if (mapIndex == null)
+                    {
+                        setBuffer[(int)index] = null;
+                        ret._nullCount++;
+                        return mapIndex;
+                    }
+                    setBuffer[(int)index] = getBuffer[mapIndex.Value];
+                    return mapIndex;
+                });
             }
+            else
+                throw new NotImplementedException(nameof(mapIndices.DataType));
```

Comment:
Should probably give a better exception and message here. Maybe an ArgumentException since `mapIndices` wasn't a valid argument.

Can a customer in the public API ever hit this? If not, maybe add an assert here.

Reply:
Public API never hits this. Added a Debug assert